Fix scrolling to pickers

jakub-trzebiatowski opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

**Based on [this PR](https://github.com/Expensify/react-native-picker-select/pull/6) to our `react-native-picker-select` fork**

1. Fixed scrolling to input components on Web
2. Added conditional ref-forwarding logic in `components/Picker`

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
3. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/14783
PROPOSAL: https://github.com/Expensify/App/issues/14783#issuecomment-1468874717


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
4. Verify a modal appears displaying a preview of that image
--->

1. Go to Settings > Workspaces > (any workspace) > Reimburse expenses > Connect bank account > Connect manually
2. Perform "Connect bank account (Step 1 of 5) anyhow (for example, fill: 011103093 / 067014822)
3. In the next step, Input correct values for all inputs before "State" picker, but leave "State" empty ("-")
4. Click/tap some text input nearby
5. Scroll the view, so the "State" field is out of sight (it should be possible on all platforms)
6. Press "Save"
7. Press "Fix the errors"

- [ ] Verify that other inputs lost focus 
  - Ensure the keyboard is hidden
  - Ensure that text inputs don't have blinking cursor
  
- [ ] Verify that the view scrolled to the picker
  - Ensure that, after scrolling, the picker ends up close to the top of the screen, consistently with other inputs 

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps

See _Tests_

<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected. **(Form storybook stories don't work for me on `main`)**
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/2590174/227771886-62fb7b3d-70e5-427d-98c6-44411c19733e.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

![image](https://user-images.githubusercontent.com/2590174/227772077-67ee6c9d-7591-4e85-a20e-1d1a3beacdc0.png)

(working as expected)

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

![image](https://user-images.githubusercontent.com/2590174/227772288-572e7c18-258b-4f94-8234-0d0edae145fe.png)

(working as expected)

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

![image](https://user-images.githubusercontent.com/2590174/227772399-1cc696ed-2b77-4792-8b59-1f97783883a7.png)

(working as expected)

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/2590174/227771903-cd44902f-92b1-499b-9dd1-47d5f3971db7.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/2590174/227771912-8422b405-af9d-4190-932b-82346ecd2c60.mp4

<!-- add screenshots or videos here -->

</details>

jakub-trzebiatowski committed:
Blur form inputs before trying to focus a new one

...as pickers and checkboxes don't really support focusing on Native.

cubuspl42 commented:
I have read the CLA Document and I hereby sign the CLA

cubuspl42 commented:
@parasharrajat I'm still going over the checklist, I'll continue tomorrow.

I tried to check relevant documents, conventions, and apply common sense.

For now, I'm wondering if we should include another form in the testing steps, as the one mentioned in the original issue ("Add debit card") doesn't really allow for scrolling on a typical screen on Android.

This form would work...:

![image](https://user-images.githubusercontent.com/2590174/227069640-cedda007-a836-425a-ae06-e8cd6039d566.png)

...but I still haven't found the path to manually trigger this flow.

cubuspl42 commented:
> * On iOS the scroll view scrolled to the picker, creating extra "stretched" void on the bottom of the screen
> 
> ![image](https://user-images.githubusercontent.com/2590174/227067637-deefec32-c101-41c2-95a3-d8ac4e4cb647.png)

@parasharrajat This is an interesting property of iOS scroll view. I personally don't have much against this effect, but we need to ensure that it is acceptable behavior.

cubuspl42 commented:
@parasharrajat Do you have thoughts on the `focusRef` naming? Pointing the `ref` to any internal widgets breaks the test suite, like I mentioned. Not obvious for me what are the work-arounds. If we'd do any pivoting, it's best to make that decision early, I think!

Also it's my first time sketching all the "Testing" sections, so any tips if I'm doing it right would be very appreciated!

cubuspl42 commented:
@parasharrajat 

> I ran the tests on all platforms & verified they passed on

Is this about the automated tests or the manual steps I describe? I'm a bit confused. Are automatic tests platform-dependent?

cubuspl42 commented:
@parasharrajat This is an example, when the user leaves out "Legal buisness name" and "State" fields in the "Company information" form on Web.

They press "Fix the errors", the form scrollst to "Legal buisness name", they fix the error, press "Fix the errors" again, then _seemingly_ the form scrolls to "Legal buisness name" **again**.

https://user-images.githubusercontent.com/2590174/227182548-36219047-e1b4-4549-ac72-a12a93b47850.mp4

It actually scrolls to "State", but focusing at the center (inconsistent with other inputs).

It's because with the current approach, we don't ensure that the `focus`/`measureLayout` contract is fulfilled, at least not on the Web. Our in-app scrolling logic is not launched, and instead we rely on the side-effect of `focus` that happens on Web, which has different behavior than the in-app logic.

I would still propose to do things consistently, also on the Web, unless we have any technical arguments against it:

https://user-images.githubusercontent.com/2590174/227190078-c830998b-2c30-49a8-8987-b21ad01de438.mp4

What would you say?

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -219,12 +219,7 @@ class Picker extends PureComponent {
-                            }
-                            this.props.innerRef(el);
-                        }}
+                        focusRef={this.props.innerRef}
```
@parasharrajat So this is the place where you would see the platform-dependent code inserted, right? Instead of `focusRef`, we'd switch whether to forward `this.props.innerRef` to `pickerProps.ref` or to `textInputProps.ref`?

cubuspl42 replied:
Okey, so I'll switch to that approach. I don't have anything against it. One more platform switch, but less changes in the fork.

jakub-trzebiatowski committed:
Form: Wrap FormSubmit in an extra View

...which is a small change with a big diff. Needed by the next commit.

jakub-trzebiatowski committed:
Form: Measure positions relative to content origin

...as measuring relative to the scroll view caused inconsistent results
between Web and Native. On Web, we basically always scrolled all the
way to the top because of this. The new approach works both on Web and
on Native.

jakub-trzebiatowski committed:
Forward the ref conditionally in components/Picker

Implement platform-specific logic that handles platform-specific
requirements and limitations.

cubuspl42 commented:
> Please Mark this PR Ready for review.

Done. But some checkboxes aren't done yet, I'll update them very soon. Also, while the PR might be ready for review (and was ready for early review before), I wouldn't say it's ready to merge until we figure out if the scrolling on Web works as expected

cubuspl42 commented:
@parasharrajat 
@johnmlee101

Hey guys, while I'm sorry that this came out this late (I spent too much time on the technicalities, including GPG and alike, it's on my side), but I'm starting to question whether the "forward ref to RNPicker's TextInput" _works at all_. While scrolling to the picker actually does work (at least on mobile, proving that TextInput's `measureLayout` works as expected), the main "selling point" of that approach was bullet-proof blur/focus behavior.

After I resolved all technical discussions and started to do all the heavy testing, I can't see **any** blur/focus behavior with this approach. I'm very uncomfortable with this, as it would mean that the original author never really tested it **or** that I can't get this to work because of some silly mistake on my side.

https://user-images.githubusercontent.com/2590174/227511679-e2d61ab7-02d3-4c2f-831e-eaaff9c31adb.mp4

It's difficult for me, as I'm working with not-fully-my proposal, but I have to take responsibility for it (as I agreed to that, and I know that).

Still, I ask you for a big favor of checking out the code, and verifying (both by reading the code and hands-on) if I implemented the "forward-to-internal -extInput" right and if you reproduce my results.

jakub-trzebiatowski committed:
Focus the input after scrolling in Form

...as on the Web it gives a slightly better visual result.

jakub-trzebiatowski committed:
Forward Picker's measureLayout to its internal TextInput

...on all platforms. This makes scrolling to pickers consistent with
other input types on Web.

cubuspl42 commented:
> I think it is fine, we agreed to not add any UI for the focused state. [#14783 (comment)](https://github.com/Expensify/App/issues/14783#issuecomment-1452972252)

What is fine? Could you elaborate?

Have you seen [the video in this post](https://github.com/Expensify/App/pull/16433#issuecomment-1482666785)?

The `focus` call on the "State" picker is a total no-op, it doesn't remove the cursor from the "Zip code" input, it doesn't hide the keyboard, it doesn't do anything.

I'm not talking about styling.

cubuspl42 commented:
> Oops. So yeah, you mentioned in your proposal to blur the inputs before focusing. Is that not working?

The "forward-to-TextInput" approach from [this proposal](https://github.com/Expensify/App/issues/14783#issuecomment-1470362844) doesn't have the blur-all block. As I understood it, it was meant to just work, re-using the TextInput shared focus. And it appears that the `focus` method in this approach is actually a **no-op** (a call doing absolutetly nothing) in case of pickers.

The only thing it does is getting a working `measureLayout` from that `TextInput`, which gives no benefits (on the technical and practical level, it's not really an opinion here) over [my proposal](https://github.com/Expensify/App/issues/14783#issuecomment-1468874717) in its actual state. Actually, it's a bit worse, as the bounds of the root `View` are _slightly_ more appropriate as a scrolling target.

I can add the blur-all block, and it's not a problem, but effectively we're just falling back to my original proposal. Do I have a green light to just do the things my way, i.e. the way that I tested, proposed, and which works? 🙂

Unless I'm missing something and there's some technical benefit of `TextInput`'s `measureLayout` over root `View`'s `measureLayout`. Please tell me if there is, we can't afford being on different pages here at this point.

cubuspl42 commented:
And I'm telling you that forwarding the ref to the `TextInput` does nothing useful. 😉 And **definitely** doesn't the thing it was meant to do, i.e. providing a working focus.

So I'll just drop that, okey? The other PR will be unnecessary.

cubuspl42 commented:
@parasharrajat 

> Does that sound good?

I would lie if I said it did. I don't understand the reasoning here, and this makes me feel uncomfortable. But I still respect your decision.

The changes as I see them [live on this branch](https://github.com/Expensify/App/compare/main...cubuspl42:App:fix-state-picker-2). They are split into a few commits, each of them doing one thing. If you'd like to review the my approach, I would recommend doing it there.

I'll try to cherry pick those changes or parts of them, so the continuity of this PR is kept.

jakub-trzebiatowski committed:
Forward Picker's focus & blur methods on Web

...for better accessibility and user experience.

This is fixing a micro-regression introduced by the previous commit.

cubuspl42 commented:
> I meant to say that we can decide on it later... First thing is to get the PR ready with all the fixed and as planned during the proposal review.

The plan derailed, from my perspective. There were assumptions behind that plan and they fell. But yup, we can go back to it in later.

I cherry picked the changes from the second branch.

@parasharrajat Commits are ready for you

cubuspl42 commented on src/components/Form.js:
```diff
@@ -307,35 +308,52 @@ class Form extends React.Component {
-                />
-                )}
-            </FormSubmit>
+            <View ref={this.formContentRef}>
```
Please note that this big diff is more readable if you read stuff commit-by-commit!

cubuspl42 replied:
@parasharrajat I didn't know that exists.

Would you expand on _why_ would I use that, if that's undocumented? I assume that it has no guarantees, and can disappear in a patch release, like all undocumented things.

cubuspl42 replied:
Have you tested this `innerViewRef` in context of the problem I'm fixing? It doesn't seem to be working.

cubuspl42 replied:
I still removed the extra view, to be double sure that we'll not introduce styling issues. At the same time, I used the inner view of `FormSubmit`, instead of `innerViewRef` from `ScrollView`, as that didn't seem to fix the problem I'm fixing.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -169,6 +166,34 @@ class Picker extends PureComponent {
+     * This method is used by Form when scrolling to the input
+     */
+    measureLayout(...args) {
+        return this.textInput.measureLayout(...args);
```
@parasharrajat This might be the best place to start the review, as without that we might focus on technicalities too early

cubuspl42 replied:
I'm resolving this thread on behalf of [that one](https://github.com/Expensify/App/pull/16433#discussion_r1148400517)

parasharrajat reviewed (changes requested):
There are a few commits that are not signed. Can you please sign all commits? 

This should help you to sign and keep the history. https://superuser.com/a/1123928

Please complete the checklist. Check off the items that do not apply to you.

Must read: 
- https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md

cubuspl42 commented on src/components/Form.js:
```diff
@@ -307,35 +308,52 @@ class Form extends React.Component {
+                        onFixTheErrorsLinkPressed={() => {
+                            // Proactively blur all inputs, as on Native we can't expect that all the inputs we'd like
+                            // to focus actually supports focusing or that their focus method blurs other inputs
+                            _.values(this.inputRefs).forEach((input) => {
+                                if (input.blur && typeof input.blur === 'function') {
+                                    input.blur();
+                                }
+                            });
```
Okey

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -169,6 +166,34 @@ class Picker extends PureComponent {
+     * This method is used by Form when scrolling to the input
+     */
+    measureLayout(...args) {
+        return this.textInput.measureLayout(...args);
```
You're right, it just doesn't. I haven't tested this newest commit on Web. I screwed up.

> Remove all focus, blur and measureLayout functions.

Please review [this thread](https://github.com/Expensify/App/pull/16433#discussion_r1147843792) on the conceptual level! 😇 Please.

I can remove and add any methods you want, but this code was like that for a reason (and I'm _not_ talking about platform switching vs platform files. I'm sorry I missed that).

cubuspl42 replied:
What would be the contract of that `getPickerRef` method? The best one I could figure out is:

```
/**
 * Returns a platform-dependent reference, which on Web exposes `focus`, but doesn't expose `measureLayout`,
 * while on Native is exposes `measureLayout`, but a non-working `focus`.
*/
```

I want to use `measureLayout` cross-platform, because it gives [**consistent results**](https://github.com/Expensify/App/pull/16433#issuecomment-1481024942). It's a user-perceived thing, not a technicality.

cubuspl42 replied:
> and to pickerProps on Web then it will have measureLayout.
Have you checked that?

cubuspl42 replied:
```js
                        pickerProps={{
                            ref: (el) => {
                                console.log({
                                    measureLayout: el.measureLayout,
                                });
                            },
```

![image](https://user-images.githubusercontent.com/2590174/227735901-0f8a56f1-0fe4-434f-beb7-51a0a0178ec2.png)

cubuspl42 replied:
I don't argue with the `Platform.OS`. It's a different topic.

In the newest push I:

- Fixed the screw-up with using `textInput` on Web (in the simplest way I could think of, i.e. by not using that reference...)
- Switched to platform files to follow our conventions

I did that by adding a boolean flag. This approach could be re-considered, i.e. maybe there could be some other way than a boolean flag. But let's leave that for later, if you'd agree.

At this point, I believe that:
* I can switch (again) to using `TextInput`, but **this will add user-perceived regression**. Scrolling will behave inconsistently for pickers and other inputs. I'm not saying **not to do that**, but I'm insisting that we talk about user-perceived effects of that **first** and then about code organization.

cubuspl42 replied:
I tried to cover this whole problem [here](https://github.com/Expensify/App/pull/16433#issuecomment-1481024942), with videos of incosistent behavior resulting from the lack of `measureLayout` on the `react-native-picker/picker`.

cubuspl42 replied:
@parasharrajat But if you don't think that that incosistency is a big deal (that we don't need/want do layout measuring on Web) I can do that picker/TextInput dispatch, I guess. I don't like the fact that it affects the user, but I also wouldn't like to miss the bonus 🙁

cubuspl42 commented:
> There are a few commits that are not signed. Can you please sign all commits?
> 
> This should help you to sign and keep the history. https://superuser.com/a/1123928
> 
> Please complete the checklist. Check off the items that do not apply to you.
> 
> Must read:
> 
>     * https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md

I read that document at the time of writing the proposal. Unfortunatelly, I didn't remember everything. I'm sorry about that.

I believed that I configured GPG right, I'll check why the non-GPG-signed commits slipped in. Sorry about that. I'll fix that.

cubuspl42 commented:
> You can configure the GPG signing for all commits in this repo. `git config commit.gpgsign true`

I did, and I also did the `exec` trick yesterday. That's why I'm not sure how the non-signed commits slipped in. But that doesn't matter. I'll be double checking now.

jakub-trzebiatowski committed:
Call Keyboard.dismiss in Form

...instead of blurring all inputs

jakub-trzebiatowski committed:
Form: Remove the extra view around FormSubmit

...and use the one inside FormSubmit instead.

This is to be totally sure that the extra view doesn't introduce styling
issues.

jakub-trzebiatowski committed:
Picker: Forward `measureLayout` calls to the root `View`

...instead of the inner text input. The reason is that the inner text
input just doesn't exist on Web.

jakub-trzebiatowski committed:
Picker: Switch to platform files

...as switching on `Platform.OS` is against our convention.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -240,4 +263,4 @@ Picker.defaultProps = defaultProps;
 
 // eslint-disable-next-line react/jsx-props-no-spreading
-export default React.forwardRef((props, ref) => <Picker {...props} innerRef={ref} key={props.inputID} />);
+export default React.forwardRef((props, ref) => <Picker {...props} ref={ref} key={props.inputID} />);
```
I'm resolving this, as this is just a consequence of a bigger approach discussed elsewhere

jakub-trzebiatowski committed:
Form: Remove the redundant innerViewRef assignment

...as these lines were committed by mistake.

cubuspl42 commented on src/components/Form.js:
```diff
@@ -312,7 +313,7 @@ class Form extends React.Component {
     render() {
         const scrollViewContent = safeAreaPaddingBottomStyle => (
-            <FormSubmit style={StyleSheet.flatten([this.props.style, safeAreaPaddingBottomStyle])} onSubmit={this.submit}>
+            <FormSubmit innerViewRef={this.formContentRef} style={StyleSheet.flatten([this.props.style, safeAreaPaddingBottomStyle])} onSubmit={this.submit}>
```
It's not. I first tested that property of `ScrollView`, and accidentaly commited both changes. I'm sorry about that and good catch.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -204,8 +235,11 @@ class Picker extends PureComponent {
+                            allowFontScaling: false,
+                        }}
                         pickerProps={{
+                            ref: this.props.focusPicker ? el => this.focusablePicker = el : undefined,
```
I started with what you're suggesting and then changed to what I commited. I don't know why. You're right, it's overcomplicated. I'll do it as you suggest.

jakub-trzebiatowski committed:
Picker: improve focus-related code

Remove the `blur` method, as it's not needed with `Keyboard.dismiss()`
being called instead of `blur` in Form (effectively removing the `blur`
method from the contract).

Improve naming and comments.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -110,9 +105,15 @@ class Picker extends PureComponent {
         this.onInputChange = this.onInputChange.bind(this);
         this.enableHighlight = this.enableHighlight.bind(this);
         this.disableHighlight = this.disableHighlight.bind(this);
+        this.blur = this.blur.bind(this);
+        this.focus = this.focus.bind(this);
+        this.measureLayout = this.measureLayout.bind(this);
```
I don't think that it's in this particular case, but my experience is that sometimes best to bind everything for consistency.

At some point a method might start getting used in a situation that needs binding, and the person doing the change might assume it's bound, becasue nearby methods were bound.

But I'm not insisting.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -110,9 +105,15 @@ class Picker extends PureComponent {
             isHighlighted: false,
         };
 
+        this.root = null;
+        this.focusablePicker = null;
+
```
It's a convention to "declare" fields in a class without the type system. While not strictly necessary, it's somewhat helpful when reading the code, I think. I'm pretty sure I saw it even in Expensify.

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -169,12 +170,42 @@ class Picker extends PureComponent {
+        }
+
+        this.focusablePicker.blur();
+    }
```
Actually I figured out `blur` is not used, if we switched to `Keyboard.dismiss()`. We effectively removed `blur` from the contract between input components and `Form`.

cubuspl42 commented:
@parasharrajat A new changeset pushed 😃

jakub-trzebiatowski committed:
Revert to the previously used react-native-picker-select version

...as the new one was required with different assumptions about fixing
the picker scrolling issues.

cubuspl42 commented on package.json:
```diff
@@ -105,7 +105,7 @@
     "react-native-performance": "^4.0.0",
     "react-native-permissions": "^3.0.1",
-    "react-native-picker-select": "git+https://github.com/Expensify/react-native-picker-select.git#84ee97dec11c2e65609511eb5a757d61bbeeab79",
+    "react-native-picker-select": "git+https://github.com/Expensify/react-native-picker-select.git#44f697f4e8be08c7ca0f697c48271e487d2a8a80",
```
Oops, right. I'm loosing control over these 15 commits. I sholuld be looking more on the PR diff 🙈

jakub-trzebiatowski committed:
Remove two unused imports

cubuspl42 commented on src/components/Form.js:
```diff
@@ -1,6 +1,6 @@
 import lodashGet from 'lodash/get';
 import React from 'react';
-import {ScrollView, StyleSheet} from 'react-native';
+import {Keyboard, ScrollView, StyleSheet, View} from 'react-native';
```
Unsed now, I'll fix it

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -1,6 +1,6 @@
 import _ from 'underscore';
 import React, {PureComponent} from 'react';
-import {View} from 'react-native';
+import {Platform, View} from 'react-native';
```
It does. I'm used to being careless in such aspects, as projects I worked on checked this on CI/CD. I'm realizing how many habits I have I should be aware of =(

jakub-trzebiatowski committed:
Form: Fix typos

cubuspl42 commented on src/components/FormSubmit/index.native.js:
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import * as formSubmitPropTypes from './formSubmitPropTypes';
 
-const FormSubmit = props => <View style={props.style}>{props.children}</View>;
+const FormSubmit = props => <View ref={props.innerViewRef} style={props.style}>{props.children}</View>;
```
While I still believe that it's mostly about specific situatio, and wheter you consider a component a wrapper over a single distinguished child ([like `FancyButton` wrapping `Button` ](https://legacy.reactjs.org/docs/forwarding-refs.html)in the official docs of `forwardRef`).

I wouldn't say this is such a case.

But at the same time, it won't cause any technical issues, so I can do that to limit extra discussions.

cubuspl42 replied:
Actually, I'm not sure how to do that without loosing props validation on the exported component or making ESLint angry on `react/prop-types` lint 😞

Actually, how would you attack this?

cubuspl42 replied:
@parasharrajat Did you have something like this in mind, or can this be improved?

```js
// ...

FormSubmit.propTypes = formSubmitPropTypes.propTypes;
FormSubmit.defaultProps = formSubmitPropTypes.defaultProps;

export default React.forwardRef((props, ref) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <FormSubmit {...props} innerRef={ref} />
));

```

cubuspl42 replied:
This disables static property validation on the exported component in some IDEs.

But I can do that, it's not the end of the world.

jakub-trzebiatowski committed:
Picker: Minor polishment

jakub-trzebiatowski committed:
Use forwardRef in FormSubmit

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -240,4 +260,4 @@ Picker.defaultProps = defaultProps;
 
 // eslint-disable-next-line react/jsx-props-no-spreading
-export default React.forwardRef((props, ref) => <Picker {...props} innerRef={ref} key={props.inputID} />);
+export default React.forwardRef((props, ref) => <Picker {...props} ref={ref} key={props.inputID} />);
```
I added one, let me know if you had that in mind or something else

cubuspl42 replied:
I went with...

```
// Forward the ref to Picker, as we implement imperative methods there
```

cubuspl42 commented on src/components/Picker/Picker.js:
```diff
@@ -169,12 +169,35 @@ class Picker extends PureComponent {
+    /**
+     * Measure the layout of this component
+     *
+     * This method is used by Form when scrolling to the input
+     */
+    measureLayout(...args) {
+        this.root.measureLayout(...args);
+    }
```
I removed the `this.<property> = null` expressions, but I didn't know what JSDoc type to use for `picker`, so I gave up on that 😏

cubuspl42 replied:
I know the syntax, I just couldn't figure out something better than `{Object}` / `{any}` type . Did you meant to use `{Object}`, or do you have some more specific type in mind?

cubuspl42 replied:
Okey, I went with that. It's better than nothing 😊

jakub-trzebiatowski committed:
Picker/index.js: Add a comment for the `ref` assignment

jakub-trzebiatowski committed:
Add JSDoc for Picker properties

jakub-trzebiatowski committed:
Form: update the usage of FormSubmit's ref

jakub-trzebiatowski committed:
Picker: remove null property assignments in the constructor

cubuspl42 commented:
@parasharrajat I believe all threads are correctly resolved. Let me know if you have any more thoughts!

I'll do some more final testing today to ensure that no regression slipped in. It would be awesome to have this merged today, as the first soft (bonus-related) deadline passes in a few hours 😀

cubuspl42 commented:
@parasharrajat Cool!

I performed manual test steps on Android, iOS and Web once again. All good.

I also ran `npm run tests`. All good.

cubuspl42 commented:
> @cubuspl42 Please merge main into this.

It has to be merge, or is rebase also okey?

cubuspl42 commented:
To be honest the size of this PR got out of hand, it's 22 commits 🙈 Is there any way to rewrite it, confirming it has the same diff as the current state?

cubuspl42 commented:
@parasharrajat Anyway, I did the merge, but please tell me if it's alright. It's the first merge in my professional career, in the companies I worked on, and same in my personal projects, I only rebased. 🙈

jakub-trzebiatowski committed:
Merge remote-tracking branch 'upstream/main' into fix-state-picker-1

cubuspl42 commented:
> It's fine.

Okey 🙃 Let me know if I can help with anything. Technically I'm past my working hours, but I'll be online for around half an hour.

jakub-trzebiatowski committed:
Picker: add JSDoc for measureLayout

jakub-trzebiatowski committed:
Rename Picker to BasePicker

jakub-trzebiatowski committed:
FormSubmit: Define the innerRef property

cubuspl42 commented:
@parasharrajat I pushed new changeset

jakub-trzebiatowski committed:
BasePicker: defer the focusing

...to work around a Mobile Safari bug.

cubuspl42 commented:
@parasharrajat Actually, I'm still not sure about how the timeline works in detail, as it's my first accepted proposal.

After the C+ review, does another "internal engineer review" start? Is there a chance to finish the whole process before the 50% penalty kicks in?

Are there any ways I, as the contributor, can help to make that happen?

cubuspl42 commented:
> For this issue, I think the timeline is already crossed. The timeline starts from the time moment you are assigned to the issue(Bot message).

We already crossed the -50% penalty deadline? That one is 6 business days, right? There should be still some time left, I think

Or are you talking about the bonus deadline? Because that one is past us, that's right. 😔 I would be grateful for any feedback on how I could have a chance for it next time

cubuspl42 commented:
@parasharrajat I'll note the feedback, thanks. From my side, I feel like some of the review rounds didn't really comment on things I introduced in the previous rounds, making it unclear why they weren't merged with the previous rounds.

But on the other hand, it might have been a bit too ambitious of me to hope for the bonus on the first issue. With the lessons learned, it might be achievable next time 😊

Cool, it's good to hear that you don't expect us to cross the penalty-causing deadline 😃

I'll be closing my workday, we can resync tomorrow.

cubuspl42 commented:
> Also, I noticed a jump in iOS Safari. Any idea what can be done about this?

I'm not even sure I reproduce it.

https://user-images.githubusercontent.com/2590174/228459931-b33929e5-366c-4607-81c2-2023a85e6bed.mp4

This looks like a slight delay with hiding the keyboard. The empty space on the bottom is just a left-over that was allocated in the layout for the keyboard. If, for debugging purposes, we modify the `Form.js` code:

```js
                <FormAlertWithSubmitButton
                    buttonText="Dismiss the keyboard"
                    // ...
                    onSubmit={() => {
                        Keyboard.dismiss();
                    }}
```

We also observe this behavior after that modification:

https://user-images.githubusercontent.com/2590174/228461434-ff67da39-0fda-4ca5-8b48-a33a62d245cd.mp4

This even happens when we press the checkbox when the keyboard is visible:

https://user-images.githubusercontent.com/2590174/228463127-7dca2577-3f70-4468-ae00-547eb364be1a.mp4

What is more, we also observe this on `main`:

https://user-images.githubusercontent.com/2590174/228464256-e34c3a94-e0cc-4264-a19e-092948717c37.mp4

cubuspl42 commented:
> Console Warning: Web:

Ouch.
So actually that `FormSubmit` prop validation doesn't work 🙁
I took a look at the console after that change, but I missed this warning.

I'll try to figure out something later today, unfortunately, I can't do this at this very moment.

cubuspl42 commented:
> Ok, let's get the. Warning fixed. I suppose we have to pass a noop function to the default value of that prop.

Like this?

```
const defaultProps = {
    innerRef: () => {},
    style: [],
};
```

Doesn't work, the warning is still present.

I'll try to figure something out.

cubuspl42 commented:
> Ok, Let me look into that. I don't want to hold this PR just for that.

While this sounds weird, in my observations the warning is non-deterministic. When I refresh the page, it sometimes is printed, and sometimes not. So I tried things that seemed to work, but after 10 more refreshes the warning returned. It doesn't help that the warning doesn't include what's wrong with `innerRef`. Printing out `ref` before it's passed like `innerRef={ref}` doesn't help; the ref looks alright, I guess?

So the thing looks a bit weird and I'm not really getting it

jakub-trzebiatowski committed:
formSubmitPropTypes.js: Change the innerRef type

jakub-trzebiatowski committed:
Merge remote-tracking branch 'upstream/main' into fix-state-picker-1

jakub-trzebiatowski committed:
BasePicker: Add more details to a comment

...regarding the Mobile Safari glitch.

parasharrajat reviewed (dismissed):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

cc: @johnmlee101 

:ribbon: :eyes: :ribbon:  C+ reviewed

johnmlee101 commented:
Lint failures

cubuspl42 commented:
> Lint failures
@johnmlee101 Ouch. Sorry about that. I'll push the fix in a few seconds.

Does the PR look good otherwise?

jakub-trzebiatowski committed:
formSubmitPropTypes.js: Remove unused import

johnmlee101 reviewed (dismissed):
I think everything looks good from my perspective. Good comments, and this should be a net positive with RBR

cubuspl42 commented:
@johnmlee101 The PR was approved by @parasharrajat before I removed the unused import. So from my perspective it's effectively approved by both of you.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4614244723) to staging by https://github.com/johnmlee101 in version: 1.2.95-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4624155945) to production by https://github.com/thienlnam in version: 1.2.95-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
